src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Cohost/CohostDocumentPullDiagnosticsEndpoint.cs

```diff
@@ -0,0 +1,204 @@
+        _logger.LogDebug($"Getting diagnostics for {razorDocument.FilePath}");
+
+        var csharpTask = GetCSharpDiagnosticsAsync(razorDocument, cancellationToken);
+        var htmlTask = GetHtmlDiagnosticsAsync(razorDocument, cancellationToken);
+
+        try
+        {
+            await Task.WhenAll(htmlTask, csharpTask).ConfigureAwait(false);
+        }
+        catch (Exception e)
+        {
+            if (e is not OperationCanceledException)
+            {
+                _logger.LogError(e, $"Exception thrown in PullDiagnostic delegation");
+                throw;
+            }
+        }
+
+        var csharpDiagnostics = await csharpTask.ConfigureAwait(false);
+        var htmlDiagnostics = await htmlTask.ConfigureAwait(false);
+
+        _logger.LogDebug($"Calling OOP with the {csharpDiagnostics.Length} C# and {htmlDiagnostics.Length} Html diagnostics");
+        var diagnostics = await _remoteServiceInvoker.TryInvokeAsync<IRemoteDiagnosticsService, ImmutableArray<LspDiagnostic>>(
+            razorDocument.Project.Solution,
+            (service, solutionInfo, cancellationToken) => service.GetDiagnosticsAsync(solutionInfo, razorDocument.Id, csharpDiagnostics, htmlDiagnostics, cancellationToken),
+            cancellationToken).ConfigureAwait(false);
+
+        if (diagnostics.IsDefaultOrEmpty)
+        {
+            return null;
+        }
+
+        _logger.LogDebug($"Reporting {diagnostics.Length} diagnostics back to the client");
+        return
+        [
+            new()
+            {
+                Diagnostics = diagnostics.ToArray(),
+                ResultId = Guid.NewGuid().ToString()
+            }
+        ];
+    }
+
+    private async Task<LspDiagnostic[]> GetCSharpDiagnosticsAsync(TextDocument razorDocument, CancellationToken cancellationToken)
+    {
+        // TODO: This code will not work when the source generator is hooked up.
+        //       How do we get the source generated C# document without OOP? Can we reverse engineer a file path?
+        var projectKey = razorDocument.Project.ToProjectKey();
+        var csharpFilePath = _filePathService.GetRazorCSharpFilePath(projectKey, razorDocument.FilePath.AssumeNotNull());
+        // We put the project Id in the generated document path, so there can only be one document
+        if (razorDocument.Project.Solution.GetDocumentIdsWithFilePath(csharpFilePath) is not [{ } generatedDocumentId] ||
+            razorDocument.Project.GetDocument(generatedDocumentId) is not { } generatedDocument)
+        {
+            return [];
+        }
+
+        _logger.LogDebug($"Getting C# diagnostics for {generatedDocument.FilePath}");
+        var csharpDiagnostics = await ExternalHandlers.Diagnostics.GetDocumentDiagnosticsAsync(generatedDocument, supportsVisualStudioExtensions: true, cancellationToken).ConfigureAwait(false);
```

Comment:
Will the work in the Razor EA to add `GetDocumentDiagnosticsAsync` work in the standalone Roslyn LSP for VS Code?

Reply:
I would expect so, yes, though all answers regarding cohosting in VS Code at this point are based solely on hopes and dreams :)